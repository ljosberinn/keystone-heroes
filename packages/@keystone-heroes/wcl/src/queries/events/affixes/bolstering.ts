import { EventDataType, HostilityType } from "../../../types";
import type { ApplyBuffEvent, RemoveBuffEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents } from "../utils";

type StackDataset = {
  targetID: number;
  targetInstance: number | null;
  stacks: number;
};

export const BOLSTERING = 209_859;

export const getBolsteringEvents = async (
  params: GetEventBaseParams
): Promise<ApplyBuffEvent[]> => {
  const allEvents = await getEvents<ApplyBuffEvent | RemoveBuffEvent>({
    ...params,
    dataType: EventDataType.Buffs,
    hostilityType: HostilityType.Enemies,
    abilityID: BOLSTERING,
  });

  const relevantEvents = allEvents.filter(
    (event): event is ApplyBuffEvent => event.type === "applybuff"
  );

  const stacksPerNPC = relevantEvents.reduce<StackDataset[]>(
    (acc, { targetID, targetInstance = null }) => {
      const existingIndex = acc.findIndex(
        (dataset) =>
          dataset.targetInstance === targetInstance &&
          targetID === dataset.targetID
      );

      if (existingIndex > -1) {
        return acc.map((dataset, index) =>
          index === existingIndex
            ? { ...dataset, stacks: dataset.stacks + 1 }
            : dataset
        );
      }

      return [
        ...acc,
        {
          targetID,
          targetInstance,
          stacks: 1,
        },
      ];
    },
    []
  );

  const highestStack = stacksPerNPC.reduce((acc, dataset) =>
    acc?.stacks >= dataset.stacks ? acc : dataset
  );

  const NPCsWithSameFinalStackAmount = stacksPerNPC.filter(
    (dataset) => dataset.stacks === highestStack.stacks
  );

  return relevantEvents.reduce<(ApplyBuffEvent & { stacks: number })[]>(
    (acc, event) => {
      const dataset = NPCsWithSameFinalStackAmount.find(
        (npc) =>
          npc.targetID === event.targetID &&
          npc.targetInstance === event.targetInstance
      );

      if (!dataset) {
        return acc;
      }

      const previousMatch = acc.find(
        (oldEvent) =>
          oldEvent.targetID === event.targetID &&
          oldEvent.targetInstance === event.targetInstance
      );

      const enhancedEvent = { ...event, stacks: dataset.stacks };

      if (!previousMatch) {
        return [...acc, enhancedEvent];
      }

      // implicitly means it must be a higher stack amount as targetID and
      // targetInstance match
      const isNewer = event.timestamp > previousMatch.timestamp;

      if (isNewer) {
        return [
          ...acc.filter((oldEvent) => oldEvent !== previousMatch),
          enhancedEvent,
        ];
      }

      return acc;
    },
    []
  );
};
