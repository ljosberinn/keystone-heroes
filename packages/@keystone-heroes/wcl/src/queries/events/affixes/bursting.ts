import { EventDataType, HostilityType } from "../../../types";
import type { DamageEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const BURSTING = 243_237;

export const getBurstingDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: BURSTING,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
