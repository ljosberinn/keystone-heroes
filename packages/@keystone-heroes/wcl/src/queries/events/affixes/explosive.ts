import { EventDataType, HostilityType } from "../../../types";
import { getEnemyNPCIDs } from "../../report";
import type { DamageEvent } from "../types";
import type { GetEventBaseParams, GetSourceIDParams } from "../utils";
import { getEvents, createEventFetcher } from "../utils";

export const EXPLOSIVE = {
  unit: 120_651,
  ability: 240_446,
};

export const getExplosiveKillEvents = async (
  params: GetEventBaseParams<{ fightID: number }>,
  explosiveID = -1
): Promise<DamageEvent[]> => {
  const sourceID =
    explosiveID === -1
      ? await getExplosiveSourceID({
          fightID: params.fightID,
          reportID: params.reportID,
        })
      : explosiveID;

  if (!sourceID) {
    return [];
  }

  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageDone,
    hostilityType: HostilityType.Friendlies,
    targetID: sourceID,
  });

  return allEvents.filter((event) => event.overkill);
};

export const getExplosiveDamageTakenEvents = createEventFetcher<DamageEvent>({
  dataType: EventDataType.DamageTaken,
  hostilityType: HostilityType.Friendlies,
  abilityID: EXPLOSIVE.ability,
});

const getExplosiveSourceID = async (params: GetSourceIDParams) => {
  const response = await getEnemyNPCIDs(
    {
      fightIDs: [params.fightID],
      reportID: params.reportID,
    },
    [EXPLOSIVE.unit]
  );

  return response[EXPLOSIVE.unit] ?? null;
};
