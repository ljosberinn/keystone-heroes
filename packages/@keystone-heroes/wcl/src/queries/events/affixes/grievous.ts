import { EventDataType, HostilityType } from "../../../types";
import type { DamageEvent } from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const GRIEVOUS_WOUND = 240_559;

export const getGrievousDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: GRIEVOUS_WOUND,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
