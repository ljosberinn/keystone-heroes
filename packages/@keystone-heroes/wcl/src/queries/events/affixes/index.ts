import { Affixes } from "@keystone-heroes/db/types";

import type { EventParams } from "..";
import type {
  AbsorbEvent,
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): (
  | AbsorbEvent
  | InterruptEvent
  | HealEvent
  | DamageEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
)[] => {
  const affixSet = new Set(affixes);

  // seasonal
  const tormentedEvents = getTormentedEvents(allEvents, affixSet);

  // common affixes
  const storming = getStormingEvents(allEvents, affixSet);
  const spiteful = getSpitefulEvents(allEvents, affixSet);
  const sanguine = getSanguineEvents(allEvents, affixSet);
  const volcanic = getVolcanicEvents(allEvents, affixSet);
  const quaking = getQuakingEvents(allEvents, affixSet);
  const bolstering = getBolsteringEvents(allEvents, affixSet);
  const bursting = getBurstingEvents(allEvents, affixSet);
  const explosive = getExplosiveEvents(allEvents, affixSet);
  const grievous = getGrievousEvents(allEvents, affixSet);
  const necrotic = getNecroticEvents(allEvents, affixSet);

  return [
    ...tormentedEvents,
    ...storming,
    ...spiteful,
    ...sanguine,
    ...volcanic,
    ...quaking,
    ...bolstering,
    ...bursting,
    ...explosive,
    ...grievous,
    ...necrotic,
  ];
};
