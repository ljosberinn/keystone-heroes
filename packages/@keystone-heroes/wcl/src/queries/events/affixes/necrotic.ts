import { EventDataType, HostilityType } from "../../../types";
import type {
  ApplyDebuffStackEvent,
  ApplyDebuffEvent,
  DamageEvent,
} from "../types";
import type { GetEventBaseParams } from "../utils";
import { getEvents, reduceEventsByPlayer } from "../utils";

export const NECROTIC = 209_858;

export const getHighestNecroticStackAmount = async (
  params: GetEventBaseParams
): Promise<ApplyDebuffStackEvent[]> => {
  const allEvents = await getEvents<ApplyDebuffStackEvent | ApplyDebuffEvent>({
    ...params,
    dataType: EventDataType.Debuffs,
    hostilityType: HostilityType.Friendlies,
    abilityID: NECROTIC,
  });

  const relevantEvents = allEvents.filter(
    (event): event is ApplyDebuffStackEvent => event.type === "applydebuffstack"
  );

  const highestStack = relevantEvents.reduce(
    (acc, event) => (acc >= event.stack ? acc : event.stack),
    0
  );

  return relevantEvents.filter((event) => event.stack === highestStack);
};

export const getNecroticDamageTakenEvents = async (
  params: GetEventBaseParams
): Promise<DamageEvent[]> => {
  const allEvents = await getEvents<DamageEvent>({
    ...params,
    dataType: EventDataType.DamageTaken,
    hostilityType: HostilityType.Friendlies,
    abilityID: NECROTIC,
  });

  return reduceEventsByPlayer(allEvents, "targetID");
};
