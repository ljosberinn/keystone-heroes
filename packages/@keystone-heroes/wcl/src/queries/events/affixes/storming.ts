import { EventDataType, HostilityType } from "../../../types";
import type { DamageEvent } from "../types";
import { createEventFetcher } from "../utils";

export const STORMING = 343_520;

export const getStormingDamageTakenEvents = createEventFetcher<DamageEvent>({
  dataType: EventDataType.DamageTaken,
  hostilityType: HostilityType.Friendlies,
  abilityID: STORMING,
});
