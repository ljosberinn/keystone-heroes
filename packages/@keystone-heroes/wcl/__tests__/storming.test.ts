import { graphql } from "msw";
import { setupServer } from "msw/node";

import { getStormingDamageTakenEvents } from "../src/queries/events/affixes/storming";
import stormingDamageTakenEvents from "./fixtures/stormingDamageTakenEvents.json";

describe("storming", () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  test("getStormingDamageTakenEvents", async () => {
    server.use(
      graphql.query("EventData", (_req, res, ctx) => {
        return res(
          ctx.data({
            reportData: {
              report: {
                events: {
                  data: stormingDamageTakenEvents,
                },
              },
            },
          })
        );
      })
    );

    const result = await getStormingDamageTakenEvents({
      reportID: "",
      startTime: 0,
      endTime: 1,
    });

    expect(result).toMatchSnapshot();
  });
});
