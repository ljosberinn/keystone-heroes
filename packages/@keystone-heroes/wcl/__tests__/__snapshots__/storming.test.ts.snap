// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`storming getStormingDamageTakenEvents 1`] = `
Array [
  Object {
    "abilityGameID": 343520,
    "absorbed": 3856,
    "amount": 2611,
    "buffs": "108366.",
    "hitType": 1,
    "mitigated": 605,
    "sourceID": -1,
    "targetID": 11,
    "timestamp": 15890552,
    "type": "damage",
    "unmitigatedAmount": 7072,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5948,
    "buffs": "24858.",
    "hitType": 1,
    "mitigated": 489,
    "sourceID": -1,
    "targetID": 122,
    "timestamp": 15890609,
    "type": "damage",
    "unmitigatedAmount": 6437,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 4383,
    "amount": 0,
    "buffs": "55342.235313.",
    "hitType": 3,
    "mitigated": 1428,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 15892388,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "sourceMarker": 6,
    "targetID": 121,
    "targetMarker": 6,
    "timestamp": 15912889,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 5480,
    "amount": 0,
    "buffs": "235313.",
    "hitType": 3,
    "mitigated": 331,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 15979100,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.203819.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "sourceMarker": 6,
    "targetID": 121,
    "targetMarker": 6,
    "timestamp": 16009965,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 382,
    "amount": 5309,
    "hitType": 1,
    "mitigated": 489,
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 360,
    "targetMarker": 2,
    "timestamp": 16018382,
    "type": "damage",
    "unmitigatedAmount": 6180,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16128563,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 382,
    "amount": 5309,
    "buffs": "108291.",
    "hitType": 1,
    "mitigated": 489,
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 360,
    "targetMarker": 2,
    "timestamp": 16158421,
    "type": "damage",
    "unmitigatedAmount": 6180,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 4816,
    "amount": 1232,
    "buffs": "108366.",
    "hitType": 1,
    "mitigated": 566,
    "sourceID": -1,
    "targetID": 11,
    "timestamp": 16169492,
    "type": "damage",
    "unmitigatedAmount": 6614,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5582,
    "hitType": 1,
    "mitigated": 480,
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 360,
    "targetMarker": 2,
    "timestamp": 16169492,
    "type": "damage",
    "unmitigatedAmount": 6062,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.203981.328900.328908.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16231101,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 5480,
    "amount": 0,
    "buffs": "235313.",
    "hitType": 3,
    "mitigated": 331,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 16231627,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 2773,
    "amount": 3814,
    "buffs": "108366.",
    "hitType": 1,
    "mitigated": 616,
    "sourceID": -1,
    "targetID": 11,
    "timestamp": 16323372,
    "type": "damage",
    "unmitigatedAmount": 7203,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 7817,
    "buffs": "337547.203819.203981.328900.187827.328908.",
    "hitType": 1,
    "mitigated": 4820,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16323919,
    "type": "damage",
    "unmitigatedAmount": 12637,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5474,
    "hitType": 1,
    "mitigated": 470,
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 360,
    "targetMarker": 2,
    "timestamp": 16324978,
    "type": "damage",
    "unmitigatedAmount": 5944,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 7817,
    "buffs": "337547.258920.203981.328900.187827.",
    "hitType": 1,
    "mitigated": 4820,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16328804,
    "type": "damage",
    "unmitigatedAmount": 12637,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5474,
    "hitType": 1,
    "mitigated": 470,
    "sourceID": -1,
    "sourceMarker": 2,
    "targetID": 360,
    "targetMarker": 2,
    "timestamp": 16338802,
    "type": "damage",
    "unmitigatedAmount": 5944,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.203819.203981.328900.328908.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16407368,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.203981.328900.328908.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16460902,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5473,
    "hitType": 1,
    "mitigated": 471,
    "sourceID": -1,
    "targetID": 360,
    "timestamp": 16464767,
    "type": "damage",
    "unmitigatedAmount": 5944,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5473,
    "hitType": 1,
    "mitigated": 471,
    "sourceID": -1,
    "targetID": 360,
    "timestamp": 16479614,
    "type": "damage",
    "unmitigatedAmount": 5944,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5135,
    "hitType": 1,
    "mitigated": 422,
    "sourceID": -1,
    "targetID": 122,
    "timestamp": 16499223,
    "type": "damage",
    "unmitigatedAmount": 5557,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 4598,
    "amount": 1390,
    "buffs": "108366.",
    "hitType": 1,
    "mitigated": 560,
    "sourceID": -1,
    "targetID": 11,
    "timestamp": 16546098,
    "type": "damage",
    "unmitigatedAmount": 6548,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 7357,
    "buffs": "337547.258920.328900.187827.102342.328908.",
    "hitType": 1,
    "mitigated": 7510,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16547186,
    "type": "damage",
    "unmitigatedAmount": 14867,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 4384,
    "amount": 0,
    "buffs": "55342.235313.",
    "hitType": 3,
    "mitigated": 1427,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 16548020,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 4247,
    "amount": 136,
    "buffs": "55342.235313.",
    "hitType": 1,
    "mitigated": 1428,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 16550539,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 1588,
    "amount": 4879,
    "buffs": "108366.",
    "hitType": 1,
    "mitigated": 605,
    "sourceID": -1,
    "targetID": 11,
    "timestamp": 16551846,
    "type": "damage",
    "unmitigatedAmount": 7072,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 4384,
    "buffs": "55342.",
    "hitType": 1,
    "mitigated": 1427,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 16553307,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 9197,
    "buffs": "337547.212084.328900.187827.",
    "hitType": 1,
    "mitigated": 5670,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16556389,
    "type": "damage",
    "unmitigatedAmount": 14867,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 2192,
    "buffs": "55342.",
    "hitType": 1,
    "mitigated": 3619,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 16558330,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6132,
    "buffs": "337547.328900.",
    "hitType": 1,
    "mitigated": 3780,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16594874,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.328900.328908.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16598710,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5419,
    "hitType": 1,
    "mitigated": 467,
    "sourceID": -1,
    "targetID": 360,
    "timestamp": 16599686,
    "type": "damage",
    "unmitigatedAmount": 5886,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.203819.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16756833,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16762579,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16788063,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5496,
    "hitType": 1,
    "mitigated": 448,
    "sourceID": -1,
    "targetID": 360,
    "timestamp": 16877415,
    "type": "damage",
    "unmitigatedAmount": 5944,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5479,
    "hitType": 1,
    "mitigated": 332,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 16882096,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5479,
    "hitType": 1,
    "mitigated": 332,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 16906184,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.203819.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16913045,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.258920.203819.203981.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16915044,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 379,
    "amount": 5988,
    "buffs": "108291.",
    "hitType": 1,
    "mitigated": 519,
    "sourceID": -1,
    "targetID": 360,
    "timestamp": 16917408,
    "type": "damage",
    "unmitigatedAmount": 6886,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6131,
    "buffs": "337547.328900.",
    "hitType": 1,
    "mitigated": 3781,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16918611,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 9197,
    "buffs": "337547.203819.328900.187827.",
    "hitType": 1,
    "mitigated": 5670,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 16929822,
    "type": "damage",
    "unmitigatedAmount": 14867,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5156,
    "buffs": "24858.",
    "hitType": 1,
    "mitigated": 401,
    "sourceID": -1,
    "targetID": 122,
    "timestamp": 16934991,
    "type": "damage",
    "unmitigatedAmount": 5557,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6158,
    "buffs": "337547.203981.328900.",
    "hitType": 1,
    "mitigated": 3754,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 17068900,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6158,
    "buffs": "337547.258920.203981.328900.",
    "hitType": 1,
    "mitigated": 3754,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 17213333,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 9237,
    "buffs": "337547.203981.328900.187827.",
    "hitType": 1,
    "mitigated": 5630,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 17371212,
    "type": "damage",
    "unmitigatedAmount": 14867,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6158,
    "buffs": "337547.258920.203981.328900.328908.",
    "hitType": 1,
    "mitigated": 3754,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 17522158,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6158,
    "buffs": "337547.258920.203981.328900.",
    "hitType": 1,
    "mitigated": 3754,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 17544504,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6158,
    "buffs": "337547.258920.203819.328900.",
    "hitType": 1,
    "mitigated": 3754,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 17675764,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5207,
    "buffs": "24858.",
    "hitType": 1,
    "mitigated": 405,
    "sourceID": -1,
    "targetID": 122,
    "timestamp": 17794517,
    "type": "damage",
    "unmitigatedAmount": 5612,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 4384,
    "amount": 0,
    "buffs": "55342.235313.",
    "hitType": 3,
    "mitigated": 1427,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 17795363,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 6158,
    "buffs": "337547.258920.328900.",
    "hitType": 1,
    "mitigated": 3754,
    "sourceID": -1,
    "targetID": 121,
    "timestamp": 17796684,
    "type": "damage",
    "unmitigatedAmount": 9912,
  },
  Object {
    "abilityGameID": 343520,
    "absorbed": 4247,
    "amount": 137,
    "buffs": "55342.235313.",
    "hitType": 1,
    "mitigated": 1427,
    "sourceID": -1,
    "targetID": 117,
    "timestamp": 17799674,
    "type": "damage",
    "unmitigatedAmount": 5811,
  },
  Object {
    "abilityGameID": 343520,
    "amount": 5206,
    "buffs": "24858.",
    "hitType": 1,
    "mitigated": 406,
    "sourceID": -1,
    "targetID": 122,
    "timestamp": 17799904,
    "type": "damage",
    "unmitigatedAmount": 5612,
  },
]
`;
